tests/test_s3_utils.py

```diff
@@ -167,3 +181,240 @@ def should_raise_any_client_error_other_than_no_such_key(
+            importer(
+                source_bucket_name=staging_bucket,
+                original_key=asset_filename,
+                target_bucket_name=storage_bucket,
+                new_key=asset_filename,
+                source_s3_client=get_s3_client_for_role(get_s3_role_arn()),
+            )
+
+            s3_etag = get_s3_etag(
+                s3_bucket=storage_bucket, s3_object_key=f"{asset_filename}", logger=get_log()
+            )
+
+            # ensure multi chunk size file
+            assert len(local_etag) >= MULTI_CHUNK_ETAG_LENGTH
+            assert local_etag == s3_etag
+        finally:
+            delete_s3_key(Resource.STORAGE_BUCKET_NAME.resource_name, asset_filename, s3_client)
+
+
+@mark.infrastructure
+def should_return_etag_if_s3_object_exists() -> None:
+    key_prefix = any_safe_file_path()
+    collection_metadata_filename = any_safe_filename()
+
+    collection_dict = {
+        **deepcopy(MINIMAL_VALID_STAC_COLLECTION_OBJECT),
+    }
+
+    with S3Object(
+        file_object=json_dict_to_file_object(collection_dict),
+        bucket_name=Resource.STORAGE_BUCKET_NAME.resource_name,
+        key=f"{key_prefix}/{collection_metadata_filename}",
+    ):
+        s3_etag = get_s3_etag(
+            Resource.STORAGE_BUCKET_NAME.resource_name,
+            f"{key_prefix}/{collection_metadata_filename}",
+            get_log(),
+        )
+
+        assert s3_etag is not None
+
+
+@patch("geostore.s3_utils.get_s3_client_for_role")
+def should_return_none_if_s3_object_is_missing(
+    get_s3_client_for_role_mock: MagicMock,
+) -> None:
+    # Given
+    bucket = any_s3_bucket_name()
+    collection_metadata_filename = any_safe_filename()
+
+    operation_name = any_operation_name()
+    error_message = any_error_message()
+    error = ClientError(
+        ClientErrorResponseTypeDef(
+            Error=ClientErrorResponseError(Code="404", Message=error_message)
+        ),
+        operation_name,
+    )
+
+    get_s3_client_for_role_mock.return_value.head_object.side_effect = error
+
+    logger_mock = MagicMock()
+
+    etag = get_s3_etag(
+        s3_bucket=bucket, s3_object_key=collection_metadata_filename, logger=logger_mock
+    )
+    assert etag is None
+
+
+@mark.infrastructure
+def should_stop_s3stacio_from_put_object_if_locally_calculated_etag_is_identical_to_s3_etag(
+    s3_client: S3Client,
+) -> None:
+
+    s3_stac_io = S3StacIO()
+    bucket = Resource.STORAGE_BUCKET_NAME.resource_name
+    collection_metadata_filename = any_safe_filename()
+
+    s3_obj_url = f"s3://{bucket}/{collection_metadata_filename}"
+    asset_contents = any_file_contents()
+
+    try:
+        # Write object to bucket - v1
+        s3_stac_io.write_text(dest=s3_obj_url, txt=str(asset_contents))
+
+        s3_response = s3_client.get_object(Bucket=bucket, Key=collection_metadata_filename)
+        obj_version_id_one = s3_response["VersionId"]
```

Comment:
Have opted for `first_write_attempt` and `second_write_attempt` instead, which may be a bit more accurate - since we are essentially making two write attempts (and second one should not happen if object already exists and etag the same). But yes, "one" and "two" is ambiguous..

Reply:
The word "version" should probably be in the name though - the current name doesn't describe the contents of the variable IMO. Maybe something (very verbose, maybe possible to shorten) like `version_id_after_nth_write_attempt`?